frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/forcepushdialog/GitForcePushDialog.java

```diff
@@ -0,0 +1,318 @@
+package com.virtuslab.gitmachete.frontend.actions.forcepushdialog;
+
+import java.awt.Component;
+import java.awt.event.ActionEvent;
+import java.util.concurrent.atomic.AtomicReference;
+
+import javax.swing.AbstractAction;
+import javax.swing.Action;
+import javax.swing.JComponent;
+import javax.swing.JPanel;
+import javax.swing.JRootPane;
+import javax.swing.border.Border;
+
+import com.intellij.dvcs.push.PrePushHandler;
+import com.intellij.dvcs.push.PushController;
+import com.intellij.dvcs.push.PushInfo;
+import com.intellij.dvcs.push.PushSource;
+import com.intellij.dvcs.push.PushSupport;
+import com.intellij.dvcs.push.PushTarget;
+import com.intellij.dvcs.push.VcsPushOptionValue;
+import com.intellij.dvcs.push.ui.PushLog;
+import com.intellij.dvcs.push.ui.VcsPushUi;
+import com.intellij.dvcs.repo.Repository;
+import com.intellij.openapi.fileEditor.FileDocumentManager;
+import com.intellij.openapi.progress.ProcessCanceledException;
+import com.intellij.openapi.progress.ProgressIndicator;
+import com.intellij.openapi.progress.Task;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.ui.DialogWrapper;
+import com.intellij.openapi.ui.Messages;
+import com.intellij.openapi.ui.ValidationInfo;
+import com.intellij.openapi.util.registry.Registry;
+import com.intellij.util.ui.JBDimension;
+import com.intellij.util.ui.JBUI;
+import com.intellij.util.ui.UIUtil;
+import com.intellij.util.ui.components.BorderLayoutPanel;
+import net.miginfocom.swing.MigLayout;
+import org.checkerframework.checker.guieffect.qual.UIEffect;
+import org.checkerframework.checker.initialization.qual.UnknownInitialization;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+public class GitForcePushDialog extends DialogWrapper implements VcsPushUi {
+  private static final String DIMENSION_KEY = "Vcs.Push.Dialog.v2";
+
+  private static final int CENTER_PANEL_HEIGHT = 450;
+  private static final int CENTER_PANEL_WIDTH = 800;
+
+  private final Project project;
+  private final PushLog listPanel;
+  private final PushController pushController;
+  private final Action pushAction;
+
+  @UIEffect
+  public GitForcePushDialog(
+      Project project,
+      java.util.Collection<? extends Repository> allRepos,
+      java.util.List<? extends Repository> selectedRepositories,
+      @Nullable Repository currentRepo,
+      @Nullable PushSource pushSource) {
+    super(project, /* canBeParent */ true,
+        (Registry.is("ide.perProjectModality")) ? IdeModalityType.PROJECT : IdeModalityType.IDE);
+    this.project = project;
+    this.pushAction = new PushSwingAction();
+    this.pushAction.putValue(DEFAULT_ACTION, Boolean.TRUE);
+
+    var dialog = new VcsPushDialogAdapterHolder(project, selectedRepositories, currentRepo, /* gitForcePushDialog */ this)
```

Comment:
Hmm how about getting rid of the holder completely + applying the following change to this file:

```diff
@@ -19,6 +19,7 @@ import com.intellij.dvcs.push.PushSupport;
 import com.intellij.dvcs.push.PushTarget;
 import com.intellij.dvcs.push.VcsPushOptionValue;
 import com.intellij.dvcs.push.ui.PushLog;
+import com.intellij.dvcs.push.ui.VcsPushDialog;
 import com.intellij.dvcs.push.ui.VcsPushUi;
 import com.intellij.dvcs.repo.Repository;
 import com.intellij.openapi.fileEditor.FileDocumentManager;
@@ -63,8 +64,7 @@ public class GitForcePushDialog extends DialogWrapper implements VcsPushUi {
     this.pushAction = new PushSwingAction();
     this.pushAction.putValue(DEFAULT_ACTION, Boolean.TRUE);
 
-    var dialog = new VcsPushDialogAdapterHolder(project, selectedRepositories, currentRepo, /* gitForcePushDialog */ this)
-        .getVcsPushDialog();
+    var dialog = new MyVcsPushDialog(project, selectedRepositories, currentRepo);
 
     this.pushController = new PushController(project, dialog, allRepos, selectedRepositories, currentRepo, pushSource);
     this.listPanel = pushController.getPushPanelLog();
@@ -76,6 +76,41 @@ public class GitForcePushDialog extends DialogWrapper implements VcsPushUi {
     init();
   }
 
+  private final class MyVcsPushDialog extends VcsPushDialog {
+
+    @UIEffect
+    MyVcsPushDialog(
+        Project project,
+        java.util.List<? extends Repository> selectedRepositories,
+        @Nullable Repository currentRepo) {
+      super(project, selectedRepositories, currentRepo);
+    }
+
+    @Override
+    @UIEffect
+    public @Nullable VcsPushOptionValue getAdditionalOptionValue(PushSupport support) {
+      return GitForcePushDialog.this.getAdditionalOptionValue(support);
+    }
+
+    @Override
+    @UIEffect
+    public void enableOkActions(boolean value) {
+      GitForcePushDialog.this.enableOkActions(value);
+    }
+
+    @Override
+    @UIEffect
+    public void updateOkActions() {
+      GitForcePushDialog.this.updateOkActions();
+    }
+
+    @Override
+    @UIEffect
+    public JRootPane getRootPane(@UnknownInitialization MyVcsPushDialog this) {
+      return GitForcePushDialog.this.getRootPane();
+    }
+  }
+
   @Nullable
   @Override
   protected Border createContentPaneBorder() {
```

Reply:
Also, maybe `enableOkActions`/`updateOkActions` could then live completely within `MyVcsPushDialog`?

Reply:
Good idea. However `updateOkActions` cannot be changes since it is called from `GitPushDialog::doValidate`.